src/rgw/rgw_flight.h

```diff
@@ -0,0 +1,222 @@
+    const DoutPrefix dp;
+    rgw::sal::Store* store;
+    FlightStore* flight_store;
+
+    std::map<std::string, Data1> data;
+
+  public:
+
+    static constexpr int default_port = 8077;
+
+    FlightServer(boost::intrusive_ptr<ceph::common::CephContext>& _cct,
+		 rgw::sal::Store* _store,
+		 FlightStore* _flight_store);
+    ~FlightServer() override;
+
+    FlightStore* get_flight_store() {
+      return flight_store;
+    }
+
+    arw::Status ListFlights(const flt::ServerCallContext& context,
+			    const flt::Criteria* criteria,
+			    std::unique_ptr<flt::FlightListing>* listings) override;
+
+    arw::Status GetFlightInfo(const flt::ServerCallContext &context,
+			      const flt::FlightDescriptor &request,
+			      std::unique_ptr<flt::FlightInfo> *info) override;
+
+    arw::Status GetSchema(const flt::ServerCallContext &context,
+			  const flt::FlightDescriptor &request,
+			  std::unique_ptr<flt::SchemaResult> *schema) override;
+
+    arw::Status DoGet(const flt::ServerCallContext &context,
+		      const flt::Ticket &request,
+		      std::unique_ptr<flt::FlightDataStream> *stream) override;
+  }; // class FlightServer
+
+  // class OwningStringView : public std::string_view {
+  class OwningStringView : public arw::util::string_view {
+
+    uint8_t* buffer;
+    int64_t capacity;
+    int64_t consumed;
+
+    OwningStringView(uint8_t* _buffer, int64_t _size) :
+      arw::util::string_view((const char*) _buffer, _size),
+      buffer(_buffer),
+      capacity(_size),
+      consumed(_size)
+    { }
+
+    OwningStringView(OwningStringView&& from, int64_t new_size) :
+      buffer(nullptr),
+      capacity(from.capacity),
+      consumed(new_size)
+    {
+      // should be impossible due to static function check
+      ceph_assertf(consumed <= capacity, "new size cannot exceed capacity");
+
+      std::swap(buffer, from.buffer);
+      from.capacity = 0;
+      from.consumed = 0;
+    }
+
+  public:
+
+    OwningStringView(OwningStringView&&) = default;
+    OwningStringView& operator=(OwningStringView&&) = default;
+
+    uint8_t* writeable_data() {
+      return buffer;
+    }
+
+    ~OwningStringView() {
+      if (buffer) {
+	delete[] buffer;
+      }
+    }
+
+    static arw::Result<OwningStringView> make(int64_t size) {
+      uint8_t* buffer = new uint8_t[size];
+      if (!buffer) {
+	return arw::Status::OutOfMemory("could not allocated buffer of size %" PRId64, size);
+      }
+      return OwningStringView(buffer, size);
+    }
+
+    static arw::Result<OwningStringView> shrink(OwningStringView&& from,
+						int64_t new_size) {
+      if (new_size > from.capacity) {
+	return arw::Status::Invalid("new size cannot exceed capacity");
+      } else {
+	return OwningStringView(std::move(from), new_size);
+      }
+    }
+
+  };
+
+  // GLOBAL
+
+  void set_flight_server(FlightServer* _server);
+  FlightServer* get_flight_server();
+  FlightStore* get_flight_store();
```

Comment:
let's avoid the globals here. i recently repurposed `struct RGWProcessEnv` in https://github.com/ceph/ceph/pull/48710 as a way to expose global state owned by `rgw::AppMain` to callers with access to `req_state`. so AppMain would own the FlightServer/Store, and the Filter could get the FlightStore pointer through `req_state::penv`